File: packages/python-bridge/connectors/github.py
"""GitHubConnector — Python mirror of ``packages/core/src/connectors/github.ts``.

The bridge process owns the ``draft`` / ``execute`` path here so it can
run without a Node sidecar. The TypeScript module is the canonical
reference for browser/Node consumers; both must agree on the wire shape
or attestations won't verify across languages.

Configuration via environment:
  - ``HMAN_GITHUB_TOKEN``       fine-grained PAT with ``issues:write``
  - ``HMAN_GITHUB_DEFAULT_OWNER`` repo owner the LLM defaults to
  - ``HMAN_GITHUB_DEFAULT_REPO``  repo name the LLM defaults to
  - ``HMAN_GITHUB_ALLOWED_REPOS``  comma-separated ``owner/repo`` whitelist
  - ``HMAN_LLM_ENDPOINT``         Ollama-compatible chat endpoint
                                  (default: http://localhost:11434/api/chat)
  - ``HMAN_LLM_MODEL``            default: ``llama3.2:3b``
"""
from __future__ import annotations

import json
import os
import re
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable, Optional

import requests

from .pact import render_attestation_block
from .types import ExecutionResult, Intention, PACTAttestation


_DRAFT_SYSTEM_PROMPT = """You are HMAN's drafting subconscious. The member casually mentioned a bug or feature request in conversation. Turn what they said into a concise, actionable GitHub issue.

Output ONLY a JSON object on a single line with two keys: "title" and "body".

Rules:
- "title" is under 70 chars, imperative mood, no trailing period.
- "body" is plain Markdown. Open with one sentence stating the problem or request from the member's point of view, then a short "## Context" section quoting their actual words verbatim.
- Don't fabricate symptoms, error messages, or stack traces the member didn't say.
- Don't add labels, milestones, or assignees - the LLM does NOT auto-categorize.
- Don't include any HMAN attestation block - that's appended separately.
"""


@dataclass
class GitHubConnectorConfig:
    default_owner: str
    default_repo: str
    allowed_repos: Optional[list[tuple[str, str]]] = None
    llm_endpoint: str = "http://localhost:11434/api/chat"
    llm_model: str = "llama3.2:3b"
    token: Optional[str] = None
    api_base: str = "https://api.github.com"

    @classmethod
    def from_env(cls) -> "GitHubConnectorConfig":
        allowed_raw = os.environ.get("HMAN_GITHUB_ALLOWED_REPOS", "").strip()
        allowed: Optional[list[tuple[str, str]]] = None
        if allowed_raw:
            allowed = []
            for entry in allowed_raw.split(","):
                entry = entry.strip()
                if "/" in entry:
                    owner, repo = entry.split("/", 1)
                    allowed.append((owner.strip(), repo.strip()))
        return cls(
            default_owner=os.environ.get("HMAN_GITHUB_DEFAULT_OWNER", "Tailor-AUS"),
            default_repo=os.environ.get(
                "HMAN_GITHUB_DEFAULT_REPO", "Human-Managed-Access-Network"
            ),
            allowed_repos=allowed,
            llm_endpoint=os.environ.get("HMAN_LLM_ENDPOINT", "http://localhost:11434/api/chat"),
            llm_model=os.environ.get("HMAN_LLM_MODEL", "llama3.2:3b"),
            token=os.environ.get("HMAN_GITHUB_TOKEN") or None,
        )


def parse_draft_reply(reply: str, context: str) -> tuple[str, str]:
    """Extract ``{title, body}`` from an LLM reply.

    Mirrors the TypeScript ``parseDraftReply`` — falls back to a stub
    issue using the raw context if parsing fails so we never silently
    drop the member's words.
    """
    # Find the first {...} JSON-shaped substring.
    start = reply.find("{")
    end = reply.rfind("}")
    if start != -1 and end != -1 and end > start:
        try:
            parsed = json.loads(reply[start : end + 1])
            title = parsed.get("title")
            body = parsed.get("body")
            if isinstance(title, str) and title.strip() and isinstance(body, str) and body.strip():
                return (title.strip()[:120], body.strip())
        except Exception:
            pass

    trimmed = re.sub(r"\s+", " ", context.strip())
    title = (trimmed[:67] + "...") if len(trimmed) > 70 else (trimmed or "Untitled issue from voice draft")
    body = (
        f"## Context\n\n> {trimmed or '(no context captured)'}\n\n"
        "_Drafted by HMAN - the LLM did not return a structured response, "
        "so this is a verbatim quote of the member's words._"
    )
    return (title, body)


class GitHubConnector:
    """Mirror of ``GitHubConnector`` (TS). ``draft`` + ``execute`` + ``undo``."""

    name = "github"

    def __init__(
        self,
        config: Optional[GitHubConnectorConfig] = None,
        *,
        llm_chat: Optional[Callable[[str, str], str]] = None,
        http_post: Optional[Callable[..., Any]] = None,
        http_patch: Optional[Callable[..., Any]] = None,
        now: Optional[Callable[[], datetime]] = None,
    ):
        self.config = config or GitHubConnectorConfig.from_env()
        self._llm_chat = llm_chat or self._default_llm_chat
        self._http_post = http_post or requests.post
        self._http_patch = http_patch or requests.patch
        self._now = now or (lambda: datetime.now(timezone.utc))

    # ── default LLM call ────────────────────────────────────────────

    def _default_llm_chat(self, system: str, user: str) -> str:
        res = requests.post(
            self.config.llm_endpoint,
            json={
                "model": self.config.llm_model,
                "messages": [
                    {"role": "system", "content": system},
                    {"role": "user", "content": user},
                ],
                "stream": False,
                "options": {"temperature": 0.2},
            },
            timeout=60,
        )
        res.raise_for_status()
        data = res.json()
        return (data.get("message") or {}).get("content", "").strip()

    # ── draft ───────────────────────────────────────────────────────

    def draft(self, *, context: str, member_id: Optional[str] = None) -> Intention:
        try:
            reply = self._llm_chat(_DRAFT_SYSTEM_PROMPT, context)
        except Exception:
            reply = ""  # fall through to stub
        title, body = parse_draft_reply(reply, context)
        # ``member_id`` reserved for future routing; not yet used in
        # default-repo selection.
        _ = member_id
        return Intention(
            id=str(uuid.uuid4()),
            connector=self.name,
            action="issue.create",
            payload={
                "owner": self.config.default_owner,
                "repo": self.config.default_repo,
                "title": title,
                "body": body,
            },
            description=f'File a GitHub issue: "{title}"',
            urgency=0.3,
            context=context,
            drafted_at=self._now().isoformat(),
        )

    # ── execute ─────────────────────────────────────────────────────

    def execute(self, intention: Intention, attestation: PACTAttestation) -> ExecutionResult:
        from .pact import hash_intention

        expected = hash_intention(intention)
        if attestation.intention_hash != expected:
            return ExecutionResult(
                success=False,
                attestation=attestation,
                error=(
                    f"attestation/intention hash mismatch "
                    f"(expected {expected}, got {attestation.intention_hash})"
                ),
            )

        owner = intention.payload.get("owner")
        repo = intention.payload.get("repo")
        title = intention.payload.get("title")
        body = intention.payload.get("body")
        if not (owner and repo and title and body):
            return ExecutionResult(
                success=False,
                attestation=attestation,
                error="payload missing owner/repo/title/body",
            )

        if self.config.allowed_repos and (owner, repo) not in self.config.allowed_repos:
            return ExecutionResult(
                success=False,
                attestation=attestation,
                error=f"repo {owner}/{repo} not in allowed_repos whitelist",
            )

        token = self.config.token
        if not token:
            return ExecutionResult(
                success=False,
                attestation=attestation,
                error=(
                    "HMAN_GITHUB_TOKEN not set - cannot call GitHub. "
                    "Use a fine-grained PAT with issues:write on the whitelisted repo."
                ),
            )

        composed_body = f"{body.strip()}\n\n{render_attestation_block(attestation)}\n"
        url = f"{self.config.api_base}/repos/{owner}/{repo}/issues"
        try:
            res = self._http_post(
                url,
                json={"title": title, "body": composed_body},
                headers={
                    "Authorization": f"Bearer {token}",
                    "Accept": "application/vnd.github+json",
                    "X-GitHub-Api-Version": "2022-11-28",
                },
                timeout=30,
            )
        except Exception as e:
            return ExecutionResult(success=False, attestation=attestation, error=str(e))

        if not getattr(res, "ok", False):
            text = getattr(res, "text", "")[:200]
            return ExecutionResult(
                success=False,
                attestation=attestation,
                error=f"GitHub issue create failed: {res.status_code} {text}",
            )

        data = res.json()
        return ExecutionResult(
            success=True,
            attestation=attestation,
            artifact_url=data.get("html_url"),
            artifact_id=f"{owner}/{repo}#{data.get('number')}",
        )

    # ── undo ────────────────────────────────────────────────────────

    def undo(self, result: ExecutionResult) -> None:
        if not result.success or not result.artifact_id:
            return
        m = re.match(r"^([^/]+)/([^#]+)#(\d+)$", result.artifact_id)
        if not m:
            raise RuntimeError(f"cannot parse artifact_id for undo: {result.artifact_id}")
        owner, repo, num = m.group(1), m.group(2), int(m.group(3))
        token = self.config.token
        if not token:
            raise RuntimeError("HMAN_GITHUB_TOKEN not set")
        headers = {
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.github+json",
            "X-GitHub-Api-Version": "2022-11-28",
        }
        # Post a "rescinded" comment, then close the issue.
        comment_url = f"{self.config.api_base}/repos/{owner}/{repo}/issues/{num}/comments"
        self._http_post(
            comment_url,
            json={
                "body": (
                    "Closed by HMAN: member rescinded consent for this action. "
                    "Original PACT attestation in the issue body remains valid history."
                )
            },
            headers=headers,
            timeout=30,
        )
        close_url = f"{self.config.api_base}/repos/{owner}/{repo}/issues/{num}"
        self._http_patch(
            close_url,
            json={"state": "closed", "state_reason": "not_planned"},
            headers=headers,
            timeout=30,
        )

File: packages/core/src/connectors/github.ts
/**
 * GitHubConnector — first concrete implementation of the Connector contract.
 *
 * Flow:
 *   1. ``draft({ context })`` — call the LLM to turn a transcript snippet
 *      into ``{ title, body }``, packaged as an Intention.
 *   2. ``execute(intention, attestation)`` — embed the attestation in a
 *      collapsed ``<details>`` block at the bottom of the issue body and
 *      POST to the GitHub REST API.
 *   3. ``undo(result)`` — close the issue with a "rescinded by member"
 *      comment.  Issues are low blast radius so undo is safe.
 */

import * as crypto from 'crypto';
import { v4 as uuidv4 } from 'uuid';
import type {
  Connector,
  ExecutionResult,
  Intention,
  PACTAttestation,
} from './Connector.js';
import {
  FetchGitHubClient,
  OllamaLLMClient,
  type GitHubClient,
  type LLMClient,
} from './types.js';

/** Payload shape the GitHub connector hands GitHub's REST API. */
export interface GitHubIssuePayload {
  owner: string;
  repo: string;
  title: string;
  body: string;
}

export interface GitHubConnectorConfig {
  /** Default repo to file against when the LLM doesn't pick one. */
  defaultOwner: string;
  defaultRepo: string;
  /** Optional whitelist — execute throws if owner/repo isn't in this list. */
  allowedRepos?: ReadonlyArray<{ owner: string; repo: string }>;
  /** Injected for testing; defaults to Ollama on localhost:11434. */
  llm?: LLMClient;
  /** Injected for testing; defaults to fetch against api.github.com. */
  github?: GitHubClient;
  /** Override clock for tests. */
  now?: () => Date;
}

/**
 * Compute the canonical hash a PACT signature should cover for an
 * Intention. Must be deterministic — every party that re-derives the
 * hash from the same Intention object gets the same value.
 */
export function hashIntention(intention: Intention<unknown>): string {
  // Stable key order so JSON.stringify is deterministic across runs.
  const stable = {
    id: intention.id,
    connector: intention.connector,
    action: intention.action,
    payload: intention.payload,
    description: intention.description,
    urgency: intention.urgency,
    context: intention.context ?? null,
    draftedAt: intention.draftedAt,
  };
  return crypto.createHash('sha256').update(JSON.stringify(stable)).digest('hex');
}

/** Render a PACT attestation as a collapsed Markdown block. */
export function renderAttestationBlock(attestation: PACTAttestation): string {
  // Pretty JSON inside the fence so a human reader can eyeball the fields.
  const json = JSON.stringify(attestation, null, 2);
  return [
    '',
    '<!-- HMAN PACT attestation — verifies this issue was an authorized member action. -->',
    '<details>',
    '<summary>HMAN PACT attestation</summary>',
    '',
    '```json',
    json,
    '```',
    '',
    '_Verify this signature with the public key above against the canonical hash of the issue payload._',
    '</details>',
  ].join('\n');
}

/**
 * Sign an Intention as a PACT attestation. The signing keypair lives
 * outside this module — pass either a base64 Ed25519 secret key plus
 * its public key, or a complete pre-built ``PACTAttestation`` shape.
 *
 * This helper exists so the bridge can call ``execute`` with the same
 * canonical attestation shape every connector embeds.
 */
export function signAttestation(input: {
  intention: Intention<unknown>;
  memberId: string;
  channel: 'voice' | 'text' | 'queue';
  privateKeyPem: string;
  publicKeyPem: string;
  timestamp?: string;
}): PACTAttestation {
  const intentionHash = hashIntention(input.intention);
  const timestamp = input.timestamp ?? new Date().toISOString();
  // Canonical bytes the signature commits to.
  const canonical = JSON.stringify({
    memberId: input.memberId,
    intentionHash,
    channel: input.channel,
    timestamp,
  });
  // Ed25519 in node 17+ uses the one-shot crypto.sign with algorithm=null
  // (Ed25519 takes the message directly, not a digest).
  const privateKey = crypto.createPrivateKey({ key: input.privateKeyPem, format: 'pem' });
  const signature = crypto.sign(null, Buffer.from(canonical, 'utf8'), privateKey).toString('base64');
  // Encode the public key as base64 of the raw PEM contents so it
  // round-trips cleanly inside JSON. Verifiers re-import with
  // ``crypto.createPublicKey({ key: pem, format: 'pem' })``.
  const publicKey = Buffer.from(input.publicKeyPem, 'utf8').toString('base64');
  return {
    memberId: input.memberId,
    intentionHash,
    channel: input.channel,
    timestamp,
    publicKey,
    signature,
  };
}

const DRAFT_SYSTEM_PROMPT = `You are HMAN's drafting subconscious. The member casually mentioned a bug or feature request in conversation. Turn what they said into a concise, actionable GitHub issue.

Output ONLY a JSON object on a single line with two keys: "title" and "body".

Rules:
- "title" is under 70 chars, imperative mood, no trailing period.
- "body" is plain Markdown. Open with one sentence stating the problem or request from the member's point of view, then a short "## Context" section quoting their actual words verbatim.
- Don't fabricate symptoms, error messages, or stack traces the member didn't say.
- Don't add labels, milestones, or assignees — the LLM does NOT auto-categorize.
- Don't include any HMAN attestation block — that's appended separately.`;

export class GitHubConnector implements Connector<GitHubIssuePayload> {
  readonly name = 'github';

  private readonly defaultOwner: string;
  private readonly defaultRepo: string;
  private readonly allowedRepos?: ReadonlyArray<{ owner: string; repo: string }>;
  private readonly llm: LLMClient;
  private readonly github: GitHubClient;
  private readonly now: () => Date;

  constructor(config: GitHubConnectorConfig) {
    this.defaultOwner = config.defaultOwner;
    this.defaultRepo = config.defaultRepo;
    this.allowedRepos = config.allowedRepos;
    this.llm = config.llm ?? new OllamaLLMClient();
    this.github = config.github ?? new FetchGitHubClient();
    this.now = config.now ?? (() => new Date());
  }

  async draft(input: { context: string; memberId?: string }): Promise<Intention<GitHubIssuePayload>> {
    const reply = await this.llm.chat({
      system: DRAFT_SYSTEM_PROMPT,
      user: input.context,
      options: { temperature: 0.2 },
    });
    const { title, body } = parseDraftReply(reply, input.context);
    const intention: Intention<GitHubIssuePayload> = {
      id: uuidv4(),
      connector: this.name,
      action: 'issue.create',
      payload: {
        owner: this.defaultOwner,
        repo: this.defaultRepo,
        title,
        body,
      },
      description: `File a GitHub issue: "${title}"`,
      // Bug reports in casual conversation are almost never urgent —
      // start low and let the member upgrade later.
      urgency: 0.3,
      context: input.context,
      draftedAt: this.now().toISOString(),
    };
    return intention;
  }

  async execute(
    intention: Intention<GitHubIssuePayload>,
    attestation: PACTAttestation,
  ): Promise<ExecutionResult> {
    // Cross-check: the attestation must commit to *this* intention.
    const expected = hashIntention(intention);
    if (attestation.intentionHash !== expected) {
      return {
        success: false,
        attestation,
        error: `attestation/intention hash mismatch (expected ${expected}, got ${attestation.intentionHash})`,
      };
    }

    const { owner, repo, title, body } = intention.payload;

    if (this.allowedRepos && !this.allowedRepos.some((r) => r.owner === owner && r.repo === repo)) {
      return {
        success: false,
        attestation,
        error: `repo ${owner}/${repo} not in allowedRepos whitelist`,
      };
    }

    const composedBody = `${body.trim()}\n\n${renderAttestationBlock(attestation)}\n`;

    try {
      const issue = await this.github.createIssue({ owner, repo, title, body: composedBody });
      return {
        success: true,
        artifactUrl: issue.html_url,
        artifactId: `${owner}/${repo}#${issue.number}`,
        attestation,
      };
    } catch (e) {
      return {
        success: false,
        attestation,
        error: e instanceof Error ? e.message : String(e),
      };
    }
  }

  async undo(result: ExecutionResult): Promise<void> {
    if (!result.success || !result.artifactId) return;
    // artifactId format: "owner/repo#number"
    const match = /^([^/]+)\/([^#]+)#(\d+)$/.exec(result.artifactId);
    if (!match) {
      throw new Error(`cannot parse artifactId for undo: ${result.artifactId}`);
    }
    const [, owner, repo, num] = match;
    await this.github.closeIssue({
      owner,
      repo,
      issue_number: Number(num),
      comment:
        'Closed by HMAN: member rescinded consent for this action. ' +
        'Original PACT attestation in the issue body remains valid history.',
    });
  }
}

/**
 * Parse the LLM's reply into ``{ title, body }``. The LLM is asked for
 * a single-line JSON object but real models leak prose around it; this
 * extracts the first JSON object and falls back to a stub-issue using
 * the raw context if parsing fails.
 */
export function parseDraftReply(
  reply: string,
  context: string,
): { title: string; body: string } {
  // Find the first { ... } JSON-shaped substring.
  const start = reply.indexOf('{');
  const end = reply.lastIndexOf('}');
  if (start !== -1 && end !== -1 && end > start) {
    const slice = reply.slice(start, end + 1);
    try {
      const parsed = JSON.parse(slice) as { title?: unknown; body?: unknown };
      const title = typeof parsed.title === 'string' && parsed.title.trim().length > 0
        ? parsed.title.trim().slice(0, 120)
        : null;
      const body = typeof parsed.body === 'string' && parsed.body.trim().length > 0
        ? parsed.body.trim()
        : null;
      if (title && body) return { title, body };
    } catch {
      // fall through to stub
    }
  }
  // Stub fallback — better to file *something* than to silently drop
  // the member's context. Title is a truncation, body quotes them.
  const trimmed = context.trim().replace(/\s+/g, ' ');
  const title = trimmed.length > 70 ? trimmed.slice(0, 67) + '...' : trimmed || 'Untitled issue from voice draft';
  const body = `## Context\n\n> ${trimmed || '(no context captured)'}\n\n_Drafted by HMAN — the LLM did not return a structured response, so this is a verbatim quote of the member's words._`;
  return { title, body };
}
